// video-react/lib/components/popup/PopupButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectSpread from "@babel/runtime/helpers/objectSpread";
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import ClickableComponent from '../ClickableComponent';
import Popup from './Popup';
var propTypes = {
  inline: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  className: PropTypes.string
};
var defaultProps = {
  inline: true
};
export default function PopupButton(props) {
  var inline = props.inline,
      className = props.className;

  var ps = _objectSpread({}, props);

  delete ps.children;
  delete ps.inline;
  delete ps.className;
  return React.createElement(ClickableComponent, _extends({
    className: classNames(className, {
      'video-react-menu-button-inline': !!inline,
      'video-react-menu-button-popup': !inline
    }, 'video-react-control video-react-button video-react-menu-button')
  }, ps), React.createElement(Popup, props));
}
PopupButton.propTypes = propTypes;
PopupButton.defaultProps = defaultProps;
PopupButton.displayName = 'PopupButton';
